### PR TITLE
feat(other): Add DragList and FlexInput

### PR DESCRIPTION
**Pull request by jbrundage**
Signed-off-by: Jaman Brundage <jbrundage372@gmail.com>

## Checklist:
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit message includes a "fixes" reference if appropriate.
  - [x] The commit is signed.
- [x] The change has been fully tested:
  - [x] I have viewed all related gallery items
  - [ ] I have viewed all related dermatology items
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised new issues to address them separately

### CODE REVIEW

**Comment by jbrundage** [comment]
1F133B01-665C-4395-9221-8B4132701161

File: packages/other/src/FlexInput.ts [hunk 1]
```diff
@@ -0,0 +1,127 @@
+import { HTMLWidget } from "@hpcc-js/common";
+
+export class FlexInput extends HTMLWidget {
+    _span;
+    _input;
+    _wrapper;
```
Comment (GordonSmith) [inline packages/other/src/FlexInput.ts]: should be `protected`

File: packages/other/src/FlexInput.ts [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+        this._span
+            .style("font-size", this.collapsedFontSize() + "px")
+            .style("top", (this.padding() / 4) + "px")
+            .style("left", this.padding() + "px")
+            ;
+        this._input
+            .style("padding", `${this.padding() * 2}px ${this.padding()}px ${this.padding() / 4}px ${this.padding()}px`)
+            .style("font-size", this.fontSize() + "px")
+            ;
+    }
```
Comment (GordonSmith) [inline packages/other/src/FlexInput.ts]: move to above `update` for consistency.

File: packages/other/src/FlexInput.ts [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+        this._span
+            .style("font-size", this.fontSize() + "px")
+            .style("top", this.padding() + "px")
+            .style("left", this.padding() + "px")
+            ;
+        this._input
+            .style("padding", `${this.padding()}px`)
+            .style("font-size", this.fontSize() + "px")
+            ;
+    }
```
Comment (GordonSmith) [inline packages/other/src/FlexInput.ts]: move to above `update` for consistency.

File: packages/other/src/FlexInput.ts [hunk 4]
```diff
@@ -0,0 +1,127 @@
...
+            .style("font-family", this.fontSize() + "px")
+            .style("transition", "all 400ms")
+            .text(this.label())
+            ;
+        this._input
+            .style("font-family", this.fontSize() + "px")
+            .style("box-sizing", "border-box")
+            .style("height", this.padding() + this.fontSize() + this.collapsedFontSize() + "px")
+            ;
+        this._input.node().value = this.value();
```
Comment (GordonSmith) [inline packages/other/src/FlexInput.ts]: Should be `this._input.prop("value", this.value())` (I think `value` is a prop and not an attr?)

File: packages/other/src/FlexInput.ts [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+
+        this._input
+            .on("focus", () => {
+                this.collapseLabel();
+            })
+            .on("blur", () => {
+                if (this._input.node().value === "") {
```
Comment (GordonSmith) [inline packages/other/src/FlexInput.ts]: SHould use `prop`

File: packages/other/src/DragList.ts [hunk 6]
```diff
@@ -0,0 +1,201 @@
...
+
+        this.widgets(orderedWidgetArr);
+        this.render();
+    }
+}
+DragList.prototype._class += " other_DragList";
+
+export interface DragList {
+    widgets(): any;
+    widgets(_: any): this;
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: Should not be "any" - should be `Widget[]`?

File: packages/other/src/DragList.ts [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+        const order = items
+            .map((elm: any) => Number(elm.getAttribute("data-order")))
+            ;
+
+        // If we iterate `order` and insert the values directly this is O(n)
+
+        // Priming an array with integers allows us to prevent a "holey" array while doing this
+        const orderedWidgetArr = [];
+        const len = this.widgets().length;
+        for (let i = 0; i < len; i++) orderedWidgetArr.push(0);
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: Change to:
```
const orderedWidgetArr = [...this.widgets()];
```

File: packages/other/src/DragList.ts [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+        const item = node.parentElement.parentElement;
+        const order = Number(item.getAttribute("data-order"));
+        this.widgets().splice(order, 1);
+        this.render();
+    }
+    applySortOrder() {
+        const items = Array.from(this._container.node().querySelectorAll(".drag-list-item"));
+        const order = items
+            .map((elm: any) => Number(elm.getAttribute("data-order")))
+            ;
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: d3.selectAll(".drag-list-item").nodes().map(...)

File: packages/other/src/DragList.ts [hunk 9]
```diff
@@ -0,0 +1,201 @@
...
+                    ;
+            })
+            ;
+        rowEnter
+            .merge(rowSelection)
+            .attr("data-order", (d, i) => i)
+            ;
+        rowSelection.exit().remove();
+    }
+    removeItem(node) {
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: needs typing

File: packages/other/src/DragList.ts [hunk 10]
```diff
@@ -0,0 +1,201 @@
...
+                    .style("height", "30px")
+                    .style("width", "30px")
+                    .style("text-align", "center")
+                    .style("line-height", "28px")
+                    .on("click", function() {
+                        context.removeItem(this);
+                    })
+                    ;
+                d.target(content.node())
+                    .render()
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: The render should really happen on the `update` and not the `enter`

File: packages/other/src/DragList.ts [hunk 11]
```diff
@@ -0,0 +1,201 @@
...
+                    ;
+            })
+            ;
+        rowEnter
+            .merge(rowSelection)
+            .attr("data-order", (d, i) => i)
+            ;
+        rowSelection.exit().remove();
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: Should set widget.target to `null` before removing it.

File: packages/other/src/DragList.ts [hunk 12]
```diff
@@ -0,0 +1,201 @@
...
+        const rowEnter = rowSelection.enter()
+            .append("div")
+            .attr("class", "drag-list-item")
+            .style("min-height", this.itemMinHeight() + "px")
+            .style("display", "flex")
+            .style("background-color", "#ddd")
+            .style("margin-bottom", this.spaceBetweenItems() + "px")
+            .style("-ms-user-select", "none")
+            .style("user-select", "none")
+            .each(function(d: any) {
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: `d` is not an any, its a Widget and you can omit the type as TS will automatically "know"

File: packages/other/src/DragList.ts [hunk 13]
```diff
@@ -0,0 +1,197 @@
+import { HTMLWidget, mouse as d3Mouse, select as d3Select, selectAll as d3SelectAll, Widget } from "@hpcc-js/common";
+
+export class DragList extends HTMLWidget {
+
+    _container;
+    _containerTop;
+    _dragMemory: any = {};
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: Should be protected

File: packages/other/src/DragList.ts [hunk 14]
```diff
@@ -0,0 +1,197 @@
...
+
+    constructor() {
+        super();
+    }
+
+    removeItem(node: HTMLElement) {
+        const item = node.parentElement.parentElement;
+        const order = Number(item.getAttribute("data-order"));
+        this.widgets().splice(order, 1);
+        this.render();
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: Should be lazyRender.

File: packages/other/src/DragList.ts [hunk 15]
```diff
@@ -0,0 +1,197 @@
...
+        const order = d3SelectAll(".drag-list-item").nodes()
+            .map((elm: any) => Number(elm.getAttribute("data-order")))
+            ;
+
+        const orderedWidgetArr = [...this.widgets()];
+
+        order.forEach((val, i) => { orderedWidgetArr[val] = this.widgets()[i]; });
+
+        this.widgets(orderedWidgetArr);
+        this.render();
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: should be lazyRender

File: packages/other/src/DragList.ts [hunk 16]
```diff
@@ -0,0 +1,197 @@
...
+    removeItem(node: HTMLElement) {
+        const item = node.parentElement.parentElement;
+        const order = Number(item.getAttribute("data-order"));
+        this.widgets().splice(order, 1);
+        this.render();
+    }
+
+    applySortOrder() {
+        const order = d3SelectAll(".drag-list-item").nodes()
+            .map((elm: any) => Number(elm.getAttribute("data-order")))
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: elm is a DIVElement?

File: packages/other/src/DragList.ts [hunk 17]
```diff
@@ -0,0 +1,197 @@
...
+            .style("position", "relative")
+            .style("overflow-x", "hidden")
+            .style("overflow-y", "auto")
+            ;
+
+        element.on("mousemove", function() {
+            if (context._dragMemory.element) {
+                const coordinates = d3Mouse(this);
+                const y = coordinates[1];
+                const parentScrollTop = context.element().node().parentElement.parentElement.scrollTop;
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: `context.element()` should be the same as `element` and `context.element().node()` should be the same as `domNode `?

File: packages/other/src/DragList.ts (hunk 11 above)
Comment (GordonSmith) [reply packages/other/src/DragList.ts]: bump

**Comment by jbrundage** [comment]
@GordonSmith these are all resolved and ready for review

File: packages/other/src/DragList.ts [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+            .map((elm: HTMLElement) => Number(elm.getAttribute("data-order")))
+            ;
+
+        const orderedWidgetArr = [...this.widgets()];
+
+        order.forEach((val, i) => { orderedWidgetArr[val] = this.widgets()[i]; });
+
+        this.widgets(orderedWidgetArr);
+        this.lazyRender();
+    }
```
Comment (GordonSmith) [inline packages/other/src/DragList.ts]: There is an easier way to do this. If the Array of "data" that is being used to generate the list is in the correct order, then you can get d3 to ensure the web items match by calling selection.order().

Comment (jbrundage) [reply packages/other/src/DragList.ts]: This method reorders the data to match the order of the dom elements. Based on the d3 documentation it seems like `selection.order()` is for doing the opposite.

Comment (GordonSmith) [reply packages/other/src/DragList.ts]: Ahh ok!

**Pull request closed without merging**